### DISCUSSION TITLE
Cancel model fetch

### DISCUSSION THREAD

**Discussion by mdsaicom** [wintercms/winter#472] [General]
Can i cancel the fetch of a model by using the afterFetch or beforeFetch events
`public function beforeFetch()
    {
        if ($this->atrribute == 'example'){
            //cancel fetch
        }
    }`

**Comment by mjauvin** [comment]
Did you try returning false ?

**Reply by mdsaicom** [reply]
Yes, but returning the model anywere

**Comment by LukeTowers** [comment]
What exactly do you mean by "cancel the fetch"? What is the expected result of that? If you're trying to prevent access to a model in the afterFetch event just throw an ApplicationException

**Reply by mdsaicom** [reply]
I need not to return the model if it does not meet a condition, but the rest of the collection if I want to return it

**Reply by LukeTowers** [reply]
What? I don't understand what you mean.

**Reply by mdsaicom** [reply]
Sorry, I'm translating with google.
I have two records in the table and of the two records I only want to return those that meet a filter at the model level but not at the query level

**Comment by mjauvin** [comment]
You mean in the List controller?

**Comment by mjauvin** [comment]
If so, read this:
https://wintercms.com/docs/backend/lists#adding-filters

And this
https://wintercms.com/docs/backend/lists#list-filters

**Reply by mdsaicom** [reply]
public function beforeFetch()
{
   $this->active = true;
   Event::fire('user.check', [&$this->active]);
   if ($this->active == false){
      //CANCEL FETCH
   }
}

**Reply by LukeTowers** [reply]
I'm still not sure what it is exactly that you're trying to do but I'm reasonably certain that is not the way to do it. Have you looked at https://wintercms.com/docs/events/event/backend.list.extendRecords?

**Comment by mdsaicom** [comment]
In frontend component this example works

$collection = MyModel::where('active', true)->get();
    
    foreach ($collection as $index => $model) {
        Event::fire('my.event', [&$model]);
        if ($model->active == false){
            unset($collection[$index]);            
        }
    }
    return $collection;
    
    -------------------------
        
    Event::listen('my.event', function (&$model) {
        if ($model->mycustomfield > 0){
            $model->active = false;
        }
    });

I need to do the same thing but in the beforeFetch event of the model

**Reply by LukeTowers** [reply]
See https://xyproblem.info/. Try asking for help with the desired end result you want rather than help with your current approach to solving that problem. I can tell you that firing an event in a loop is a pretty bad idea for performance reasons so that's already a huge negative to your proposed approach; and secondly attempting to exclude a model from a collection by doing something in the afterFetch event is not going to do what you want it too either.

**Comment by mdsaicom** [comment]
Ok. Thank you for your answers.
Finally, I'm going to choose to use collection overloading if you don't tell me otherwise.

    /**
     * Create a new Collection instance.
     */
    public function newCollection(array $models = [])
    {
        return new CustomCollection($models);
    }

**Reply by LukeTowers** [reply]
I'm pretty sure that's not the best way to go about it but since you haven't provided enough information on your use case for me to provide better alternatives if that works for you then go for it.
